### ISSUE TITLE
restartlessness and e10s

### ISSUE THREAD

**Issue by myrdd** [RequestPolicyContinued/requestpolicy#489]
The aim is to to

1. convert RP into a [restartless extension](https://developer.mozilla.org/en-US/Add-ons/Bootstrapped_extensions) (aka *bootstrapped*)
2. support multiprocess firefox (aka *Electrolysis* or *e10s*)

**Development branch: [`dev-restartless-and-e10s`](https://github.com/RequestPolicyContinued/requestpolicy/tree/dev-restartless-and-e10s)**

As of 2014-12-02 the main work is already done, so the aim is to **find and remove all remaining bugs**.

- [x] implement restartlessness
- [x] implement e10s support
- [ ] testing and finding bugs – [known issues](https://github.com/RequestPolicyContinued/requestpolicy/issues?q=is%3Aissue+is%3Aopen+[restartless%2Fe10s]+label%3Adev-branch)

~~IMHO it's great if you don't need to restart firefox when you install/disable/enable the addon. But the main reason I want to do is that overlay extensions are not supported by the [Add-on Debugger](https://developer.mozilla.org/en-US/Add-ons/Add-on_Debugger), which is built-in since ff 31.~~

~~I think it should be possible. A tutorial how to convert can be found [here](https://developer.mozilla.org/en-US/Add-ons/How_to_convert_an_overlay_extension_to_restartless). As RP also implements an XPCOM component (nsIContentPolicy), we need to consider [this](http://stackoverflow.com/questions/18354668/how-to-implement-xpcom-component-nsicontentpolicy-in-bootstrapped-firefox-exte).~~

**Comment by myrdd** [comment]
There are still several issues. I'm not documenting them online, but I'm working on fixing them.

**Comment by myrdd** [comment]
To show you I'm still there and working on the problems, I've pushed a bunch of commits to `dev-restartless-and-e10s`, see [here](https://github.com/RequestPolicyContinued/requestpolicy/commits/dev-restartless-and-e10s).

The main remaining issue is that RP does not cleanly shut down yet, i.e. when „disable“ is clicked on `about:addons`.

**Comment by myrdd** [comment]
Sadly the restartless/e10s brach is still too buggy to be merged into `dev-1.0`. I'm still working on the startup/shutdown process working cleanly. After that I'll fix the bugs that have been reported – thank you @nysatrok and @ldgbc for testing and reporting the issues! :+1:

**Comment by myrdd** [comment]
> I'm still working on the startup/shutdown process working cleanly.

I think this is done, at least I don't get any errors when I enable/disable the addon.
- The commit I've just pushed is 75aca763efa379a7a51d229246bf928ce79aa67a.
- I've added some documentation about the [„Environment System“](https://github.com/RequestPolicyContinued/requestpolicy/wiki#the-environment-system). I think it's too text heavy, so some day I might add some diagrams.

Now it's time to fix the more obvious bugs that have been reported.

**Comment by myrdd** [comment]
My current plan is to merge the [`dev-restartless-and-e10s`](https://github.com/RequestPolicyContinued/requestpolicy/tree/dev-restartless-and-e10s) branch as soon as it works well when E10s is **disabled**. *Real* E10s support is then another task, but currently there are more important issues, which is the reason why I want to merge that branch back to `dev-1.0`.

**Comment by myrdd** [comment]
I've just pushed some more commits.

Sorry for being silent recently, it's examination period. As finishing this issue is still the next step before releasing RPC to AMO, it would be great if you @ldgbc @nodiscc could do some testing if you can bring up the time. Other testers are also welcome; creating the XPI should be fairly easy.

**Comment by nodiscc** [comment]
@myrdd could you release a `dev-restartless-and-e10s` XPI at https://github.com/RequestPolicyContinued/requestpolicy/releases (and explicitely mark it as "development version for testers)? I don't have access to my dev/build machine right now.

**Comment by nodiscc** [comment]
@myrdd time for a new release? :)

**Comment by myrdd** [comment]
@nodiscc working on it. I want to make a better workaround for #447 for that release.

**Comment by ldgbc** [comment]
Not sure if there much testing to be done. I don't see any problem with it except for one that I'll report after this post.

Question though, is the https://github.com/RequestPolicyContinued/requestpolicy/tree/dev-restartless-and-e10s consider dead now? I notice the merge to dev1.0 built and it more updated by commit counts.

@nodiscc you could just make an XPI using any software that can create a ZIP files, that how I build RP at least.

**Comment by myrdd** [comment]
You are right @ldgbc, I've now removed the `dev-restartless-and-e10s` branch.

> @nodiscc you could just make an XPI using any software that can create a ZIP files, that how I build RP at least.

That could work, but might not work anymore in future. I've introduced a preprocessor, which is used for example in the Logger, see [here](https://github.com/RequestPolicyContinued/requestpolicy/commit/c7358990da4025c5b878520e2bb4f29065eac547#diff-4d578029105218550c4d6f2b115b5144).
